datafusion/expr/src/async_udf.rs

```diff
@@ -0,0 +1,135 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::{ScalarFunctionArgs, ScalarUDF, ScalarUDFImpl};
+use arrow::array::ArrayRef;
+use arrow::datatypes::{DataType, SchemaRef};
+use async_trait::async_trait;
+use datafusion_common::config::ConfigOptions;
+use datafusion_common::error::Result;
+use datafusion_common::internal_err;
+use datafusion_expr_common::columnar_value::ColumnarValue;
+use datafusion_expr_common::signature::Signature;
+use std::any::Any;
+use std::fmt::{Debug, Display};
+use std::sync::Arc;
+
+/// A scalar UDF that can invoke using async methods
+///
+/// Note this is less efficient than the ScalarUDFImpl, but it can be used
+/// to register remote functions in the context.
+///
+/// The name is chosen to mirror ScalarUDFImpl
+#[async_trait]
+pub trait AsyncScalarUDFImpl: Debug + Send + Sync {
+    /// the function cast as any
+    fn as_any(&self) -> &dyn Any;
+
+    /// The name of the function
+    fn name(&self) -> &str;
+
+    /// The signature of the function
+    fn signature(&self) -> &Signature;
+
+    /// The return type of the function
+    fn return_type(&self, _arg_types: &[DataType]) -> Result<DataType>;
```

Comment:
Should this trait also define `return_field_from_args` to match `ScalarUDFImpl` and to allow extension types in async UDFs?

Reply:
Thanks for mentioning this point. I didn't notice that the method of `ScalarUDFImpl` was updated. Yes, I think it's better to follow the definition of `ScalarUDFImpl`. I'll update it.